### DISCUSSION TITLE
Make it possible to cancel panel queries

### DISCUSSION THREAD

**Discussion by sunker** [grafana/grafana#40191] [Plugins]
**What would you like to be added?**:

I would like to make it possible to cancel panel queries from within the query editor row. 

**Why?**

For many cloud databases, such as Redshift and Athena, the customer pay per query execution time. For BI related databases, it's not unusual that queries take minutes and possibly hours to run. Yet it might be hard to know in advance how long a query will actually take. Since this costs moneys, it's crucial to have the ability to cancel the long running queries. 

---

It is already possible to cancel the query in the panel chrome. However, it's not completely obvious to all user that this is possible. User has to hoover the spinner to see the `Cancel query` tooltip. This will cancel the query context, which should stop the running query on the upstream given context cancellation is handled in the backend. 

The ideal user experience would be to have a run button and a cancel button on each query row. This is an example of the Athena query editor in the AWS console. 
![Screenshot from 2021-10-08 08-13-22](https://user-images.githubusercontent.com/2388950/136513940-f2af3802-0448-490e-b536-11e77f24c072.png)
This would include plugin api changes since it's currently not possible to run `one` query only for a certain panel. 

Another option would be to make the existing panel query cancellation easier to discover for users. Possibly by adding a button i the query group header. This would be an improvement compared to how it is today, but still not ideal.

**Comment by mckn** [comment]
I really like this idea and it makes sense in the context that you describe. 
Two solutions that comes to my mind:

**We extend the datasource plugin props with the capabilities of canceling request**
- The drawback with this is that you then need to implement this kind of controls in all datasources that would like this behavior.
- The benefit would be that you have full control in the data source query editor on how the cancelation would work.

**We extend the datasource plugin registration API with a value where you can enable/disable these controls**
If they are enabled we will display `run`/`cancel`/`time` controls in the query editor row. One tricky thing here is to find a good way to add it the UI. Maybe @Ijin08 can help out here?
 
- The benefit is that we only need to implement this once and it can be enabled/disabled on a datasource plugin basis. 
- The drawback is that if the query editor would not have the flexibility to implement this in a datasource specific way.

**Reply by mckn** [reply]
Another option would be to expose both ways of doing this.

**Reply by sunker** [reply]
Yeah definitely a question for UX, but after giving it an extra thought I think it might be hard to satisfy all needs using the query editor row header. Might be better to leave that up to query editor developers to decide where to put the `run/cancel/time` buttons. 

If we ought to have a run/cancel button on the query editor row, it should only run that certain query. In that case, I think this needs to be an extension of the query editor props.
```typescript
   ...
  onRunQuery: () => void; // The as-is - runs all queries in the panel
  onRunQueryRow: (refId: string) // Runs one certain query in the panel. Not sure if this is feasible, but this func could possibly return a 
  // promise so that the query editor can handle the cancellation and reflect loading state in the run/cancel button etc.   
``` 
The onRunQueryRow would simple run datasource.query with only the specified refId. I guess we need to think carefully about this though since it would mean a pretty substantial change to how query editors behave. 

onRunQueryRow is a terrible name :)

**Reply by mckn** [reply]
> Yeah definitely a question for UX, but after giving it an extra thought I think it might be hard to satisfy all needs using the query editor row header. Might be better to leave that up to query editor developers to decide where to put the run/cancel/time buttons.

Yeah - maybe a good idea to start with implementing this in the query editor. If we feel that we are starting to repeat ourself in multiple query editors we can move the implementation to core Grafana so it can be shared between multiple editor.

Regarding the `onRunQueryRow` I think that would work. hmm, maybe it would make sense to do something like this:

```ts
...
onRunQuery: () => void;
getQueryContext: (refId: string) => QueryContext;
```

```ts
type QueryContext = {
   onRun: () => void;
   onCancel: () => void;
};
```

Then we also could provide this via a context aware hook as well.

**Reply by mckn** [reply]
Also, we would have the possibility to add more things to the `QueryContext`

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
